js/tji-datasets-explore.js

```diff
@@ -390,17 +447,64 @@ TJIChartView.prototype.filter_data = function() {
     return true;
   })
 
-  this.update_charts(data);
+  this.state.filtered_data = data;
 }
 
-TJIChartView.prototype.update_charts = function(data) {
-  this.update_record_count(data);
+TJIChartView.prototype.update_charts = function() {
+  var that = this;
   _.each(this.state.charts, function(chart){
-    chart.update(data);
+    chart.update(that.state.filtered_data);
   })
 }
 
-TJIChartView.prototype.update_record_count = function(data) {
+TJIChartView.prototype.update_record_count = function() {
   if (this.state.$record_count) jQuery(this.state.$record_count).remove();
-  this.state.$record_count = jQuery(this.record_count_template.replace('{count}', data.length)).prependTo(this.charts_elt_id);
+  this.state.$record_count = jQuery(this.record_count_template.replace('{count}', this.state.filtered_data.length)).prependTo(this.state.$above_charts_area);
+}
+
+TJIChartView.prototype.download = function() {
+  // Download all the data the user is currently viewing (that is,
+  // with their current filters applied).
+
+  // Note: This gets tricky because we want to download
+  // not only the fields available in the charts, but ALL
+  // fields (columns) in the complete dataset.
+
+  // Make note of the ids of our current (filtered) records.
+  var ids_to_keep = {}
+  var that = this;
+  _.each(this.state.filtered_data, function(record) {
+    ids_to_keep[record[that.state.record_id_field]] = true;
+  });
+
+  // Pull out the matching complete-dataset records.
+  var filtered_complete_records = _.filter(this.state.complete_data, function(record) {
+    return ids_to_keep[record[that.state.record_id_field]] !== undefined;
+  });
+  console.log("Downloading " + filtered_complete_records.length + " records (of " + this.state.data.length + ")");
+  
+  // Convert these JSON records into CSV text using Papa
+  var csvData = Papa.unparse(filtered_complete_records);
+  var filename = filtered_complete_records.length == this.state.complete_data.length ? "tji_data.csv" : "tji_data_filtered.csv";
```

Comment:
could you timestamp the filename or some such, so they don't accidentally overwrite prior to download, if they tryna download multiple filtered data sets. 
Also is there some way we could include what filters were selected -- do csvs have meta data or sumpin? P sure not o/w you woulda included it.

Reply:
Most browsers will append something to downloads that have a filename clash. On my mac, it'll do `tji_data.csv` then `tji_data (1).csv`.

As for filters... I think this needs additional thought. It's a good idea in theory, but the filenames could get nasty long if we want to fully specify all filters that are applied. Eh?